### DISCUSSION TITLE
Older iOS versions support

### DISCUSSION THREAD

**Discussion by rogerfar** [rhashimoto/wa-sqlite#112] [Q&A]
Because WebSQL is deprecated and actually being [removed](https://developer.chrome.com/blog/deprecating-web-sql/) we're looking for a replacement. Currently I'm testing wa-sqlite with an indexeddb backend, primarily because we need to support iOS.

When testing, I found that it works fine on iOS until 15.4, anything lower it throws the following error:
```
WebAssembly.Module doesn't parse at byte 117: invalid opcode 193, in function at index 78). Build with -sASSERTIONS for more info. (evaluating 'new WebAssembly.RuntimeError(e+". Build with -sASSERTIONS for more info.")
```

This seems to come from some flags that aren't supported in the older wasm runtimes, found some tips here: https://github.com/emscripten-core/emscripten/issues/15673 and https://github.com/emscripten-core/emscripten/issues/17609

I'm wanting to recompile sqlite to be more compatible, but are there any implications that it could have down the line with the DBBatchAtmoicVFS?

**Comment by rhashimoto** [comment]
A problem I know you would run into is locking. All the example VFS classes that support multiple connections (e.g. from multiple browser tabs), including IDBBatchAtomicVFS, use [WebLocks to implement locking](https://github.com/rhashimoto/wa-sqlite/blob/master/src/examples/WebLocks.js). Safari added [WebLocks support in 15.4](https://caniuse.com/?search=lockmanager), so if you want to go back further than that then you'll need to replace the locking implementation.

If you don't need to support SQLite's locking mechanism - i.e. if there will never be multiple connections wanting to access the database simultaneously, or if there are but your application manages the arbitration before either one gets to call SQLite - then it's a trivial patch to just return SQLITE_OK from the VFS methods xLock/xUnlock (edit: actually the base class does this, so you could just delete the subclass methods). Otherwise you can look through the history of [this FAQ item](https://github.com/rhashimoto/wa-sqlite/issues/5) to see alternatives to WebLocks that I considered, but I'll warn you that everything before WebLocks is annoying and painful. I think I concluded that a service worker was the best alternative, which really says something.

I don't _think_ there are any other browser APIs required that need 15.4, but I can't guarantee that. It's also quite possible that I use some Javascript language features that might have arrived in Safari relatively recently - at least that could be handled by transpiling.
